### PR TITLE
Grid primitive

### PR DESCRIPTION
**Pull request by reesscot**
*Description of changes:*

This PR implements the Grid primitive. The `rowSpan` and `columnSpan` props don't map directly to CSS Properties, and thus need to be transformed to `grid-row: span ${spanValue}/span ${spanValue}` and `grid-column: span ${spanValue}/span ${spanValue}` CSS properties. The rowSpan and columnSpan props work can be responseive when the customer specifies styles like this:

```jsx
    <Grid
      templateColumns="300px 300px"
      autoRows="100px"
      justifyContent="center"
      gap="20px">
      <Card rowSpan={3}>1</Card> {/** <-- look here **/}
      <Card columnSpan={2}>2</Card>  {/** <-- look here **/}
    </Grid>
```

Or  using the array or object syntax:

```jsx
    <Grid 
      templateColumns="300px 300px"
      autoRows="100px"
      justifyContent="center"
      gap="20px">
      <Card rowSpan={[1,2]}>1</Card> {/** <-- look here **/}
      <Card columnSpan={{base: 1, large: 2}}>2</Card> {/** <-- look here **/}
    </Grid>
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/react/__tests__/exports.ts [hunk 1]
```diff
@@ -1866,23 +1868,16 @@ describe('@aws-amplify/ui-react', () => {
           "VisuallyHidden",
           "Wrapper",
           "components",
-          "convertStylePropsToStyleObj",
```
Comment (reesscot) [inline packages/react/__tests__/exports.ts:1869]: Removed these util functions, because I don't see any reason for them to be exposed on the public api.

File: packages/react/src/primitives/shared/__tests__/utils.test.tsx [hunk 2]
```diff
@@ -30,119 +20,6 @@ const props: ViewProps = {
   className: 'my-section',
 };
 
-let breakpoints = theme.breakpoints.values;
```
Comment (reesscot) [inline packages/react/src/primitives/shared/__tests__/utils.test.tsx:33]: Moved these out to styleUtils.ts file

File: packages/react/src/primitives/types/style.ts [hunk 3]
```diff
@@ -16,7 +17,7 @@ export type ResponsiveStyle<PropertyType> =
   | PropertyType[]
   | ResponsiveObject<PropertyType>;
 
-export interface BaseStyleProps {
+export interface BaseStyleProps extends GridItemStyleProps {
```
Comment (reesscot) [inline packages/react/src/primitives/types/style.ts:20]: This allows all primitives to be used as Grid items

File: packages/react/src/primitives/utils/testUtils.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+export const errorMessageWrapper = (fn: () => void, message: string) => {
```
Comment (reesscot) [inline packages/react/src/primitives/utils/testUtils.ts:1]: This function allows us to render a customer error message for an expect that throws.

File: packages/react/__tests__/exports.ts (hunk 1 above)
Comment (hvergara) [reply packages/react/__tests__/exports.ts:1869]: thanks!

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
+import * as React from 'react';
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:1]: any reason to use `*` instead of default export?

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+import autoprefixer from 'autoprefixer';
+import postcssJs from 'postcss-js';
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:3]: nit: `postcssJs` => `postcss`

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  GridItemStyleProps,
+  ResponsiveObject,
+  ViewProps,
+} from '../types/index';
+
+import { getValueAtCurrentBreakpoint } from './responsive/utils';
+import { useBreakpoint } from './responsive/useBreakpoint';
+import { Breakpoint, Breakpoints } from '../types/responsive';
+
+import { useTheming } from '../../theming/';
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:19]:
```suggestion
import { useTheming } from '../../theming';
```

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+export const prefixer = postcssJs.sync([autoprefixer]);
+
+import {
+  ComponentPropsToStylePropsMap,
+  ComponentPropsToStylePropsMapKeys,
+  GridItemStyleProps,
+  ResponsiveObject,
+  ViewProps,
+} from '../types/index';
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:13]:
```suggestion
} from '../types';
```

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+import { filterOutNullOrEmptyStringValues, isNullOrEmptyString } from './utils';
+
+/**
+ * Transforms style props to another target prop
+ * where the original is a simpler API than the target.
+ * This function will remove the original prop and
+ * replace target prop values with calculated
+ * E.g. rowSpan => row, columnSpan => column
+ * @param props T
+ * @returns
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:29]: are these from another doc?

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  }, [rowSpan, columnSpan, convertGridSpan]);
+
+  return {
+    row: !isNullOrEmptyString(row) ? row : rowFromSpanValue,
+    column: !isNullOrEmptyString(column) ? column : columnFromSpanValue,
+    ...rest,
+  };
+};
+
+export const usePropStyles = (props: ViewProps, style: any) => {
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:50]:
```suggestion
export const usePropStyles = (props: ViewProps, style: React.CSSProperties) => {
```

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+    if (Array.isArray(spanValue)) {
+      return spanValue.map((value) => getGridSpan(value));
+    } else {
+      const newObj: ResponsiveObject<string> = {};
+      Object.entries(spanValue).forEach(([key, value]) => {
+        newObj[key] = getGridSpan(value);
+      });
+      return newObj;
+    }
+  }
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:117]:
```suggestion
  if (Array.isArray(spanValue)) {
      return spanValue.map(getGridSpan);
  }
  
  if (typeof spanValue === 'object' && spanValue != null) {
    const newObj: ResponsiveObject<string> = {};
    Object.entries(spanValue).forEach(([key, value]) => {
      newObj[key] = getGridSpan(value);
    });
    return newObj;
  }
```

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+export const convertStylePropsToStyleObj: ConvertStylePropsToStyleObj = ({
+  props = {},
+  style = {},
+  breakpoint,
+  breakpoints,
+}) => {
+  ComponentPropsToStylePropsMapKeys.filter((stylePropKey) =>
+    filterOutNullOrEmptyStringValues(props[stylePropKey])
+  ).forEach((stylePropKey) => {
+    let value = getValueAtCurrentBreakpoint(
```
Comment (hvergara) [inline packages/react/src/primitives/shared/styleUtils.ts:141]: if `value` never change, can be a `const`?

File: packages/react/src/primitives/shared/utils.ts [hunk 13]
```diff
@@ -1,116 +1,21 @@
...
-  } = useTheming();
+export const isFn = (fn: unknown): fn is Function => typeof fn === 'function';
 
-  const breakpoint = useBreakpoint({
-    breakpoints,
-    breakpointUnit,
-    defaultBreakpoint: defaultBreakpoint as Breakpoint,
-  });
+export const isNullOrEmptyString = (value: unknown) =>
+  value == null || (typeof value === 'string' && !strHasLength(value));
```
Comment (hvergara) [inline packages/react/src/primitives/shared/utils.ts:14]: `strHasLength` already checks for value type (`string`)

```suggestion
  value == null || !strHasLength(value);
```

File: packages/react/src/primitives/types/flex.ts [hunk 14]
```diff
@@ -1,34 +1,14 @@
...
-   */
-  alignItems?: ResponsiveStyle<Property.AlignItems>;
-
-  /**
-   * Sets the distribution of space between and around content items
-   */
-  alignContent?: ResponsiveStyle<Property.AlignContent>;
-
+import { CSSLayoutStyleProps, ResponsiveStyle } from './style';
+export interface FlexStyleProps extends CSSLayoutStyleProps {
```
Comment (hvergara) [inline packages/react/src/primitives/types/flex.ts:5]:
```suggestion

export interface FlexStyleProps extends CSSLayoutStyleProps {
```

File: packages/react/src/primitives/types/style.ts (hunk 3 above)
Comment (hvergara) [reply packages/react/src/primitives/types/style.ts:20]: great!

File: docs/src/components/DemoBox.tsx [hunk 15]
```diff
@@ -1,17 +1,53 @@
...
   primitiveName: string;
+  startCollapsed?: boolean;
 }
 
 export const DemoBox: React.FC<DemoBoxProps> = ({
   primitiveName,
   children,
+  startCollapsed = true,
 }) => {
+  const [collapsed, toggleCollapsed] = React.useState(startCollapsed);
```
[the remark this answers is not in the document]
Comment (reesscot) [reply docs/src/components/DemoBox.tsx:19]: good point, will update

File: docs/src/pages/ui/primitives/grid/demo.tsx [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+          alignItems={gridContainerProps.alignItems}
+          autoColumns={gridContainerProps.autoColumns}
+          autoRows={gridContainerProps.autoRows}
+          columnGap={gridContainerProps.columnGap}
+          gap={gridContainerProps.gap}
+          rowGap={gridContainerProps.rowGap}
+          templateAreas={gridContainerProps.templateAreas}
+          templateColumns={gridContainerProps.templateColumns}
+          templateRows={gridContainerProps.templateRows}
+          justifyContent={gridContainerProps.justifyContent}
```
[the remark this answers is not in the document]
Comment (reesscot) [reply docs/src/pages/ui/primitives/grid/demo.tsx:96]: I had this originally, but then React complains when we try to pass the setter functions to the DOM element. e.g., `setJustifyContent` is not a value DOM attribute

File: packages/react/src/primitives/shared/styleUtils.ts (hunk 5 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/styleUtils.ts:1]: Yes, this is the new recommended way to import React. See:
https://epicreact.dev/importing-react-through-the-ages/

TLDR (my understanding): React doesn't actually expose a default export. We've been able to do this previously by using the [esModuleInterop](https://www.typescriptlang.org/tsconfig#esModuleInterop) setting which enables [allowSyntheticDefaultImports](https://www.typescriptlang.org/tsconfig#allowSyntheticDefaultImports), which converts `import React from 'react'` to `import * as React from 'react'`. Doing a namespace import is ESM compatible and future proof.



https://twitter.com/housecor/status/1386839524362407943?lang=en

File: packages/react/src/primitives/shared/styleUtils.ts (hunk 6 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/styleUtils.ts:3]: updated

File: packages/react/src/primitives/shared/styleUtils.ts (hunk 9 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/styleUtils.ts:29]: This was a mistake. I've removed it since that information is in the Types

File: packages/react/src/primitives/shared/styleUtils.ts (hunk 11 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/styleUtils.ts:117]: Good suggestion, will make the change

File: packages/react/src/primitives/shared/styleUtils.ts (hunk 12 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/styleUtils.ts:141]: Good catch, will update

File: packages/react/src/primitives/shared/utils.ts (hunk 13 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/utils.ts:14]: Good point, updated.

**Review by zchenwei — CHANGES_REQUESTED** [review]
Thanks for all the work! This is looking great and just left a few comments.

File: packages/react/src/primitives/types/grid.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import { Property } from 'csstype';
+import { AriaProps, BaseComponentProps } from './base';
+
```
[the remark this answers is not in the document]
Comment (reesscot) [reply packages/react/src/primitives/types/grid.ts:3]: I assume you mean the spacing here. Fixed.

File: packages/react/src/primitives/shared/utils.ts [hunk 18]
```diff
@@ -1,116 +1,21 @@
...
-      prefixer(
-        convertStylePropsToStyleObj({ props, style, breakpoint, breakpoints })
-      ),
-    [props, style, breakpoints, breakpoint]
-  );
-};
-
-const filterOutNullOrEmptyStringValues = (value) =>
+export const filterOutNullOrEmptyStringValues = (value) =>
   value != null && (typeof value !== 'string' || strHasLength(value));
```
Comment (hvergara) [inline packages/react/src/primitives/shared/utils.ts:17]: I personally think we can rid of this function (as we can use `isNullOrEmptyString` as replacement). But, if you don't agree, we can simplify it a lot, just returning the opposite of `isNullOrEmptyString`:
 
```suggestion
export const filterOutNullOrEmptyStringValues = (value) => !isNullOrEmptyString(value);
```

File: packages/react/src/primitives/shared/utils.ts [hunk 19]
```diff
@@ -1,116 +1,21 @@
...
-  const {
-    theme: {
-      breakpoints: {
-        values: breakpoints,
-        unit: breakpointUnit,
-        defaultBreakpoint,
-      },
-    },
-  } = useTheming();
+export const isFn = (fn: unknown): fn is Function => typeof fn === 'function';
```
Comment (hvergara) [inline packages/react/src/primitives/shared/utils.ts:11]: minor: can we call it `isFunction`? Also, I couldn't find any reference to this typeguard in our codebase. Why do we need this?

File: packages/react/src/primitives/shared/styleUtils.ts (hunk 5 above)
Comment (hvergara) [reply packages/react/src/primitives/shared/styleUtils.ts:1]: discussed offline: we'll change default exports in a separate PR

File: docs/src/pages/ui/primitives/grid/demo.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (hvergara) [reply docs/src/pages/ui/primitives/grid/demo.tsx:96]: got it, thanks!

**Review by hvergara — APPROVED** [review]
Minor comments about some shared util functions (not a blocker though). Feel free to change it or not, I'm approving anyway :)

File: packages/react/src/primitives/shared/utils.ts (hunk 19 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/utils.ts:11]: Yes, will update the name. We are using it in some new PR's that haven't been merged yet.

File: packages/react/src/primitives/types/grid.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  columnSpan?: ResponsiveStyle<number | 'auto'>;
+  columnStart?: ResponsiveStyle<Property.GridColumnStart>;
+  row?: ResponsiveStyle<Property.GridRow>;
+  rowEnd?: ResponsiveStyle<Property.GridRowEnd>;
+  rowSpan?: ResponsiveStyle<number | 'auto'>;
+  rowStart?: ResponsiveStyle<Property.GridRowStart>;
+}
+
+export interface GridProps
+  extends BaseComponentProps,
```
[the remark this answers is not in the document]
Comment (reesscot) [reply packages/react/src/primitives/types/grid.ts:30]: Fixed, and updated to extend ViewProps

File: packages/react/src/primitives/shared/styleUtils.ts [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    });
+    return newObj;
+  }
+  return null;
+};
+
+export const getGridSpan = (spanValue: number | 'auto'): string => {
+  return spanValue === 'auto'
+    ? 'auto'
+    : `span ${spanValue} / span ${spanValue}`;
```
[the remark this answers is not in the document]
Comment (reesscot) [reply packages/react/src/primitives/shared/styleUtils.ts:114]: Fixed

File: packages/react/src/primitives/shared/utils.ts (hunk 18 above)
Comment (reesscot) [reply packages/react/src/primitives/shared/utils.ts]: Good idea, removed.

**Review by zchenwei — APPROVED** [review]
LGTM!
